File: csharp/1090.largest-values-from-labels.cs
/*
 * @lc app=leetcode id=1090 lang=csharp
 *
 * [1090] Largest Values From Labels
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int LargestValsFromLabels(int[] values, int[] labels, int num_wanted, int use_limit) {
        return values
            .Zip(labels, (int value, int label) => (Label: label, Value: value))
            .OrderByDescending(t => t.Value)
            .GroupBy(t => t.Label, (int label, IEnumerable<(int Label, int Value)> all) => all.Take(use_limit).Select(t => t.Value))
            .SelectMany(all => all)
            .OrderByDescending(x => x)
            .Take(num_wanted)
            .Sum();
    }
}
// @lc code=end

File: java/1090.largest-values-from-labels.java
/*
 * @lc app=leetcode id=1090 lang=java
 *
 * [1090] Largest Values From Labels
 */

import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Queue;

// @lc code=start

import static java.util.Comparator.comparing;

class Solution {
    public int largestValsFromLabels(int[] values, int[] labels, int numWanted, int useLimit) {
        Map<Integer, Integer> count = new HashMap<>();
        int n = values.length;
        Queue<Pair> queue = new PriorityQueue<>(comparing(Pair::value).reversed());
        for (int i = 0; i < n; ++i) {
            queue.offer(new Pair(values[i], labels[i]));
        }
        int ret = 0;
        while (numWanted > 0 && !queue.isEmpty()) {
            Pair p = queue.poll();
            int used = count.getOrDefault(p.label(), 0);
            if (used == useLimit) {
                continue;
            }
            --numWanted;
            count.put(p.label(), used + 1);
            ret += p.value();
        }
        return ret;
    }

    private record Pair(int value, int label) {}
}
// @lc code=end

